feat: add IDepositTradeHelper and Uniswap packages

ramenforbreakfast opened the pull request:
Added some pre-emptive dev comments around `depositAndTrade` to provide context on function inputs.

XavierEkkel commented on apps/smart-contracts/core/package.json:
```diff
@@ -48,6 +48,8 @@
         "@openzeppelin/contracts-upgradeable": "4.7.3",
         "@typechain/ethers-v5": "8.0.3",
         "@typechain/hardhat": "3.0.0",
+        "@uniswap/v3-core": "1.0.1",
```
will we need core as well? the interface only uses periphery

ramenforbreakfast replied:
We need the core because the `ISwapRouter` interface inherits from a contract in `v3-core`

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/IDepositTradeHelper.sol:
```diff
@@ -0,0 +1,33 @@
+
+interface IDepositTradeHelper {
+  /**
+   * @dev `baseTokenAmount` will be taken from user and deposited to mint
+   * preUSD. This minted preUSD will be used entirely towards purchasing
+   * either Long or Short tokens from the corresponding UniswapV3 pool.
```
```suggestion
   * @dev `baseTokenAmount` will be taken from msg.sender (assumes that msg.sender has already approved at least `baseTokenAmount` of the Collateral's base token) and then deposited to mint
   * Collateral. This minted Collateral will be used entirely towards purchasing
   * either Long or Short tokens from the corresponding UniswapV3 pool.
```

XavierEkkel replied:
what value will be used for minAmountOut in the swap router?

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/IDepositTradeHelper.sol:
```diff
@@ -0,0 +1,33 @@
+   */
+  function depositAndTrade(
+    uint256 baseTokenAmount,
+    IPrePOMarket market,
+    bool long
```
why not just pass in the relevant pool directly instead? the relevant pool address can be retrieved off-chain rather than on-chain

ramenforbreakfast commented:
I have updated this based on our discussion here => https://www.notion.so/prepo/DepositTradeHelper-SC-Spec-37ec0e75d82f402d8c68ac49e1562a7e

Detailed descriptions of structs will be added here based on the information in this doc. Struct variable naming follows the naming convention used in the functions they will be used with (`ERC20Permit` and `SwapRouter`) to prevent confusion.

XavierEkkel reviewed (commented):
needs updating to reflect latest spec

ramenforbreakfast commented:
Updated with `getBaseToken` and removed immutable setters. Was going to upgrade to OZ 4.8.0 thinking they had removed the draft naming, but still has yet to be released.

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/IDepositTradeHelper.sol:
```diff
@@ -0,0 +1,50 @@
+  struct OffChainTradeParams {
+    address tokenOut;
+    address deadline;
+    address amountOutMinimum;
```
```suggestion
    uint256 amountOutMinimum;
```

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/IDepositTradeHelper.sol:
```diff
@@ -0,0 +1,50 @@
+
+  function getCollateral() external view returns (ICollateral);
+
+  function getBaseToken() external view returns (IERC20Permit);
```
nit: move above `getCollateral` to have consistent order with `depositAndTrade` arguments

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/IDepositTradeHelper.sol:
```diff
@@ -0,0 +1,50 @@
+    address amountOutMinimum;
+    uint160 sqrtPriceLimitX96;
+  }
+
```
missing `POOL_FEE_TIER` constant

XavierEkkel reviewed (approved):
LGTM after comments
